### PR TITLE
Gen modification

### LINKED ISSUE #13: Modify event generator
Modifications made are done in order to create a stress test in order to monitor the performance when visual improvements are made.

Acceptance criteria:

- Roughly 30 000 events are sent an hour
- Randomised time between events sent
- Links to be random (if applicable)

### PR DESCRIPTION
**Pull request by gusaxtcha**
# Generator Modification 

## Description: 
  * Added Eiffel event list to the generator to allow for random event types
  * Changed the maximum amount of links per event to allow for a more realistic graph 
  * Added CLI arguments to allow for control over runtime and latency between events sent  
    * Arguments added: 

  | Argument name| Short name | Description | Default value|
  | ------------- | ------------- | ------------- | ------------- |
  | total_duration  | -t  | Total amount of time in milliseconds for the generator to run  |  3 600 000   |
  | min_latency  | -m  | Minimum value for the range of latency between events sent   |  5  |
  | latency_max  | -l  | Maximum value for the range of latency between events sent   |  220  |

## Useful notes:
* The default values of the CLI arguments (count, total_duration, min_latency and latency max ) are set to mimic roughly 30 000 events set per hour.
* The generator will run until either the count (total events to be sent) or the total_duration (duration for generator to run) has been reached.

## Screenshots:
Preview of random event types with random links:

Screenshot 2022-10-14 195140

closes #13

### CODE REVIEW

**Comment by RuthgerD** [comment]
Hard coding event names here is not how the library is designed.

Please look into how the event name was derived originally before your change.

**Comment by gusaxtcha** [comment]
> Hard coding event names here is not how the library is designed.
> 
> Please look into how the event name was derived originally before your change.

I will Look into it!

**Comment by gusaxtcha** [comment]
@RuthgerD I have looked into the issue you raised with the hard coded values. Following your suggestion lead me to the event_set.rs file, specifically the impl Event function. Am I on the right track here? I figured it was cleaner to ask before implementing something that might not be following the intended design.

**Comment by RuthgerD** [comment]
> @RuthgerD I have looked into the issue you raised with the hard coded values. Following your suggestion lead me to the event_set.rs file, specifically the impl Event function. Am I on the right track here? I figured it was cleaner to ask before implementing something that might not be following the intended design.

Yes EventSet is used to describe what events and links are generated, this also allows us to have different link types per event if wish to do so. It is already used in the event sender (EventSet::build), so you should really extend that instead of modifying the internals. I haven't looked at your changes too closely but I think just the existing EventSet will get you what you want already.

**Comment by gusaxtcha** [comment]
> > @RuthgerD I have looked into the issue you raised with the hard coded values. Following your suggestion lead me to the event_set.rs file, specifically the impl Event function. Am I on the right track here? I figured it was cleaner to ask before implementing something that might not be following the intended design.
> 
> Yes EventSet is used to describe what events and links are generated, this also allows us to have different link types per event if wish to do so. It is already used in the event sender (EventSet::build), so you should really extend that instead of modifying the internals. I haven't looked at your changes too closely but I think just the existing EventSet will get you what you want already.

Thanks for your guidance! I ended up following your advice having the implementation in the event sender itself. I have changed the list of events to a global constant (I figured for such a small list and the fact that the generator is used for testing, implementing something to read from txt/json or config file would be a bit of an overkill for the intended usage).

**Comment by RuthgerD** [comment]
Could you show how your EventSet looks?

**Comment by gusaxtcha** [comment]
> Could you show how your EventSet looks?

It's the same as in the PR, I reverted back and made the changes in the event  sender.

**Comment by RuthgerD** [comment]
Ok so what we have currently is:
```rust
EventSet::build()
  .add_link(Link::new("Link0", true))
  .add_link(Link::new("Link1", true))
  .add_event(
     Event::new("Event", "1.0.0")
        .with_link("Link0")
        .with_link("Link1"),
   )
  .build()
  .expect("This should work")
 ```
 What is prevent you from adding more events like
 ```rust
EventSet::build()
  .add_link(Link::new("Link0", true))
  .add_link(Link::new("Link1", true))
  .add_link(Link::new("Link2", true))
  .add_event(
     Event::new("Event1", "1.0.0")
        .with_link("Link0")
        .with_link("Link1"),
   )
  .add_event(
     Event::new("Event2", "1.0.0")
        .with_link("Link0")
   )
  .add_event(
     Event::new("Event3", "1.0.0")
        .with_link("Link2"),
   )
  .build()
  .expect("This should work")
 ```
 The generator should then pick one randomly if constraints are met (https://github.com/ItJustWorksTM/EiffelVis/blob/master/backend/crates/eiffelvis_gen/src/generator.rs#L137)

**Comment by gusaxtcha** [comment]
@RuthgerD I actually tried that initially and thought it might be a bit messy but if you are happy with it like that, I will implement it that way but use less types to keep in neat.
Thanks again for your help!

**Comment by RuthgerD** [comment]
> @RuthgerD I actually tried that initially and thought it might be a bit messy but if you are happy with it like that, I will implement it that way but use less types to keep in neat. Thanks again for your help!

You can loop over the types you want to add and just call builder = builder.add_event(...); if you think it would be too many lines.

Theres always a way to unmessy the mess

File: backend/tools/event_sender/src/main.rs
Comment (RuthgerD) [inline backend/tools/event_sender/src/main.rs:148]: This should be inlined to just `rand::thread_rng().gen_range(from..to)` as there is only one use of it

File: backend/tools/event_sender/src/main.rs
Comment (RuthgerD) [inline backend/tools/event_sender/src/main.rs:166]: this while loop is inside the `for _ in 0..(target)` loop, even if it works it is not correct logic

Comment (gusaxtcha) [reply backend/tools/event_sender/src/main.rs:166]: Good point! Will update!

File: backend/tools/event_sender/src/main.rs
Comment (gusaxtcha) [reply backend/tools/event_sender/src/main.rs:148]: We were using it for something else too initially but will update.

File: backend/tools/event_sender/src/main.rs
Comment (gusaxtcha) [reply backend/tools/event_sender/src/main.rs:166]: Of course, not sure why I have placed it within the for loop! Will update. Thanks!

**Comment by RuthgerD** [comment]
Code seems OK now, can't vouch for the actual functionality though

**Comment by gusaxtcha** [comment]
> Code seems OK now, can't vouch for the actual functionality though

Great. It's been tested thoroughly and will be tested again independently by the assigned reviewer. Any issues will be fixed before the PR is accepted.

**Comment by RuthgerD** [comment]
This is related to: https://github.com/ItJustWorksTM/EiffelVis/issues/14, only event types at the moment though, would be cool to also have link types match the official ones.

**Comment by gusaxtcha** [comment]
> This is related to: [ItJustWorksTM#14](https://github.com/ItJustWorksTM/EiffelVis/issues/14), only event types at the moment though, would be cool to also have link types match the official ones.

We can definitely look into that! It will probably be done on a new branch though.

**Comment by RuthgerD** [comment]
> > This is related to: [ItJustWorksTM#14](https://github.com/ItJustWorksTM/EiffelVis/issues/14), only event types at the moment though, would be cool to also have link types match the official ones.
> 
> We can definitely look into that! It will probably be done on a new branch though.

Sounds good, anyways would like me to approve the PR (for the future as well?) we're still on your repo :p

**Comment by alrapal** [comment]
It's always good to have your approval @RuthgerD ! 
one of us (me it seems) will review this as well! But at least quality check has passed :D Thanks for you investment, we appreciate it!

**Comment by gusaxtcha** [comment]
> > > This is related to: [ItJustWorksTM#14](https://github.com/ItJustWorksTM/EiffelVis/issues/14), only event types at the moment though, would be cool to also have link types match the official ones.
> > 
> > 
> > We can definitely look into that! It will probably be done on a new branch though.
> 
> Sounds good, anyways would like me to approve the PR (for the future as well?) we're still on your repo :p

Our way of working currently is that the reviewer will approve the PR but if you (or someone else on your side) are going to review our changes anyway, we can always chat about changing the required number of approvals to 2 and one of them can be someone from your side? :)

**Comment by RuthgerD** [comment]
Hmm let's just treat our reviews as secondary, we will have to review stuff again upstream anyways.

**Comment by gusaxtcha** [comment]
> Hmm let's just treat our reviews as secondary, we will have to review stuff again upstream anyways.

Sounds good! Thanks for all your input on this one!

**Review by alrapal — CHANGES_REQUESTED** [review]
Code looks good and is commented, well done

Some exploratory testing results here:

## Running without parameters: ✅ 
 Everything seems to work on both BE - FE
Screenshot 2022-10-24 at 13 42 58

## Running with -h  ✅ 
Explanations are clear 
Screenshot 2022-10-24 at 13 43 44

## Running with parameters 
### -- replay.  ✅ 
It seems to work on its own.

### - l and - m are equal: 🙅 
Problem when those are equal, it makes the thread panic. 
Screenshot 2022-10-24 at 14 03 47

### -c & -b alone and with --replay ✅ 
Seems all good! 

For the only bug I found, it should be an easy fix by branching out of the random if they are the same. I suspect it comes from there

Nice work otherwise!

**Comment by gusaxtcha** [comment]
> Code looks good and is commented, well done
> 
> Some exploratory testing results here:
> 
> ## Running without parameters: ✅
> Everything seems to work on both BE - FE Screenshot 2022-10-24 at 13 42 58
> 
> ## Running with -h ✅
> Explanations are clear Screenshot 2022-10-24 at 13 43 44
> 
> ## Running with parameters
> ### -- replay. ✅
> It seems to work on its own.
> 
> ### - l and - m are equal: 🙅
> Problem when those are equal, it makes the thread panic. Screenshot 2022-10-24 at 14 03 47
> 
> ### -c & -b alone and with --replay ✅
> Seems all good!
> 
> For the only bug I found, it should be an easy fix by branching out of the random if they are the same. I suspect it comes from there
> 
> Nice work otherwise!

Nice catch with the latency! Thanks for that! Changes have been made. :)

**Review by alrapal — APPROVED** [review]
Nice with the if greater! 
I did not think about this :D 
Good job!

**Pull request merged**
